pennylane/transforms/core/transform_program.py

```diff
@@ -131,10 +131,14 @@ def __bool__(self):
         return bool(self._transform_program)
 
     def __add__(self, other):
-        program = TransformProgram(self._transform_program)
-        for container in other:
-            program.push_back(container)
-        return program
+        if self.has_final_transform and other.has_final_transform:
+            raise TransformError("The transform program already has a terminal transform.")
+
+        transforms = self._transform_program + other._transform_program
+        if self.has_final_transform:
+            transforms.append(transforms.pop(len(self) - 1))
```

Comment:
I get that this is the only way to make addition work if `self.has_final_transform`, but I fear that this might not be the most natural/assumed behaviour. I think I would prefer to raise a warning here letting the user know explicitly what they've done, but I'm curious for your thoughts. cc @trbromley

Reply:
I don't think a warning is the best idea. The perfect example for why this behaviour should happen is device transforms. If a user applies an transforms to a QNode, the order of the transforms in the transform program of that QNode should be `User transforms -> device transforms -> terminal transforms`. I should also note that `TransformProgram.push_back` does raise an error if a user tries applying a transform after a terminal transform. `__add__` is for internal use only.